Implement `scalaImports` trait

lenguyenthanh opened the pull request:
This is to implement `scalaImports` trait, which provides mechanism to add an import to a specific generated `structure` from smithy. The rationale behind is on this [discussion](https://github.com/disneystreaming/smithy4s/discussions/1525).

I still need to check all the items below but want to make a draft PR first to gather some early feedback.

ps: You can see this pr in action here: https://github.com/lenguyenthanh/fide/pull/100

## PR Checklist (not all items are relevant to all PRs)

- [x] Added unit-tests (for runtime code)
- [x] Added bootstrapped code + smoke tests (when the rendering logic is modified)
- ~~[ ] Added build-plugins integration tests (when reflection loading is required at codegen-time)~~
- ~~[ ] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)~~
- ~~[ ] Updated dynamic module to match generated-code behaviour~~
- [x] Added documentation
- [x] Updated changelog

lenguyenthanh committed:
Add scalaImports trait and a short documentation

lenguyenthanh committed:
Implement ScalaImportsTrait

lenguyenthanh committed:
Implement ScalaImports for Hint ADT

This also update SmithyToIR to convert ScalaImportsTrait pojo
to Hint ADT

lenguyenthanh committed:
Extract imports from scalaImports trait and render them

lenguyenthanh commented on modules/protocol/resources/META-INF/smithy/smithy4s.meta.smithy:
```diff
@@ -163,3 +163,14 @@ structure generateOptics {}
+/// to compile. espically in the case you want to compose refinement types
+/// and other validators. `providerImport` should be an import that the
+/// target shape is required to compile.
+@trait(selector: "structure :not([trait|error])") // todo remove trait
```
For now, I only limit it to `structure`, where I guess it makes most sense (and make it easier to implement). But maybe I should open it to `union` and `newtype` as well.

Baccata replied:
You can lift the constraint altogether imho

lenguyenthanh replied:
do you mean it should just be: `@trait(selector: "*")`?

Baccata replied:
I mean it should just be `@trait`, no selector needed.

lenguyenthanh replied:
fixed in https://github.com/disneystreaming/smithy4s/pull/1550/commits/5873cc511a505f720e01c7e9de6b812a03feee9a

Baccata commented on modules/protocol/resources/META-INF/smithy/smithy4s.meta.smithy:
```diff
@@ -163,3 +163,14 @@ structure generateOptics {}
+/// and other validators. `providerImport` should be an import that the
+/// target shape is required to compile.
+@trait(selector: "structure :not([trait|error])") // todo remove trait
+structure scalaImports {
```
I'd make it a 

```smithy
list scalaImports {
  member: Import
} 
```

lenguyenthanh replied:
fixed in https://github.com/disneystreaming/smithy4s/pull/1550/commits/627af24f1b96596d0b9a8c0fe30458974c43f392 after some struggle with java collections

lenguyenthanh committed:
Change scalaImports from structure to list

lenguyenthanh committed:
Add test for scalaImport renderer

Baccata commented:
Looks good to me. The only other thing that I'd add would be some smoke tests, namely adding some specs in `sampleSpecs` folder that will be generated during the compilation of the `bootstrapped` module. 

If you can reproduce a minimal spec showcasing the problem that led you to issue this PR, and how it can be fixed with this new trait, I'll deem the PR mergeable 👍

lenguyenthanh commented:
> Looks good to me. The only other thing that I'd add would be some smoke tests, namely adding some specs in `sampleSpecs` folder that will be generated during the compilation of the `bootstrapped` module.

Thanks, I was looking at how to do smoke tests, but couldn't find. I'll add it today.


> If you can reproduce a minimal spec showcasing the problem that led you to issue this PR, and how it can be fixed with this new trait, I'll deem the PR mergeable 👍

Should I write this in to a docs? probably `04-codegen/01-customisation/15-scala-imports.md`?

lenguyenthanh committed:
Remove all constrains from scalaImports

lenguyenthanh committed:
Add more tests for scalaImports renderer

This also fix union renderer issue

lenguyenthanh committed:
Add mixing refinement and scalaImports

lenguyenthanh committed:
Add minimal smoke test

lenguyenthanh commented:
Added a minimal smoke test for this. And it also shows We can see why this We need this feature.

```smithy
@scalaImports(["smithy4s.refined.Age.provider._"])
structure StructureWithScalaImports {
  @range(min: 13, max: 19)
  teenage: Age
}
```
This wouldn't compiled without the needed provider(which is delivered by `scalaImports` trait):
```scala
    implicit val rangeProvider: RefinementProvider.Simple[smithy.api.Range, Age] =
      RefinementProvider.rangeConstraint(x => x.value)
```

lenguyenthanh commented:
So, I think this is done code wise, now times for some documentation and change log. Is this okay better to do those in another PR and let this one going first?

Baccata commented:
@lenguyenthanh I'd rather the docs were added to this PR, to be honest. 

> Should I write this in to a docs? probably 04-codegen/01-customisation/15-scala-imports.md? 

This is a fine place, yes 👍

lenguyenthanh commented:
cool, I'll add it tomorrow.

lenguyenthanh committed:
Add documentation for scalaImports

lenguyenthanh committed:
Update CHANGELOG

lenguyenthanh commented:
Added documentation, and updated changelog. Tested by building the website locally. Please help me improve those docs 🙏 

![Screenshot 2024-05-31 at 12-28-31 Add Scala imports to generated code smithy4s](https://github.com/disneystreaming/smithy4s/assets/437967/6538901b-eec1-46e5-bd5c-ce84d22ffa84)

Baccata reviewed (approved):
The documentation is great, thank you very much !

lenguyenthanh commented:
thanks for your guidelines @Baccata . It was fun!
